fcrepo-kernel-impl/src/test/java/org/fcrepo/kernel/impl/services/CreateResourceServiceImplTest.java

```diff
@@ -0,0 +1,264 @@
+        verify(psSession).persist(ArgumentMatchers.any(ResourceOperation.class));
+    }
+
+    @Test
+    public void testWithSlugExistsRdf() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        when(psSession.getHeaders(fedoraId + "/" + "testSlug", null)).thenReturn(resourceHeaders);
+        when(resourceHeaders.getTypes()).thenReturn(basicContainerTypes);
+        createResourceService.perform(txId, fedoraId, "testSlug", null, null, model);
+        verify(psSession).persist(ArgumentMatchers.any(ResourceOperation.class));
+    }
+
+    @Test
+    public void testWithSlugDoesntExistsRdf() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        when(psSession.getHeaders(fedoraId + "/" + "testSlug", null))
+                .thenThrow(PersistentItemNotFoundException.class);
+        when(resourceHeaders.getTypes()).thenReturn(basicContainerTypes);
+        createResourceService.perform(txId, fedoraId, "testSlug", null, null, model);
+        verify(psSession).persist(ArgumentMatchers.any(ResourceOperation.class));
+    }
+
+    @Test(expected = ItemNotFoundException.class)
+    public void testNoParentBinary() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenThrow(PersistentItemNotFoundException.class);
+        createResourceService.perform(txId, fedoraId, null, null, null, digests,
+                null, null);
+    }
+
+    @Test(expected = CannotCreateResourceException.class)
+    public void testParentIsBinary() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(resourceHeaders.getTypes()).thenReturn(nonRdfSourceTypes);
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        createResourceService.perform(txId, fedoraId, null, null, null, digests,
+                null, null);
+    }
+
+    @Test
+    public void testSlugIsNullBinary() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        when(resourceHeaders.getTypes()).thenReturn(basicContainerTypes);
+        createResourceService.perform(txId, fedoraId, null, null, null, digests,
+                null, null);
+        verify(psSession).persist(ArgumentMatchers.any(ResourceOperation.class));
+    }
+
+    @Test
+    public void testWithSlugExistsBinary() throws Exception {
+        final String fedoraId = UUID.randomUUID().toString();
+        when(psSession.getHeaders(fedoraId, null)).thenReturn(resourceHeaders);
+        when(psSession.getHeaders(fedoraId + "/" + "testSlug", null)).thenReturn(resourceHeaders);
```

Comment:
Might make sense to use the helper method added in this PR for joining these strings